File: src/memento/control/operations.py
from __future__ import annotations

import hashlib
import json
import sqlite3
from dataclasses import dataclass
from datetime import UTC, datetime
from enum import StrEnum


class OperationState(StrEnum):
    QUEUED = "queued"
    RUNNING = "running"
    SUCCEEDED = "succeeded"
    FAILED = "failed"
    CONFLICT = "conflict"
    RECOVERING = "recovering"


class IdempotencyConflictError(RuntimeError):
    """Raised when the same principal/key pair is reused for another request."""


@dataclass(frozen=True, slots=True)
class OperationRequest:
    op_id: str
    principal: str
    idempotency_key: str
    tool_name: str
    request_json: str
    client_instance_id: str | None = None
    mcp_session_id: str | None = None
    source_chat: str | None = None

    @property
    def request_hash(self) -> str:
        return hashlib.sha256(self.request_json.encode("utf-8")).hexdigest()


@dataclass(frozen=True, slots=True)
class OperationRecord:
    op_id: str
    principal: str
    idempotency_key: str
    tool_name: str
    request_hash: str
    state: OperationState
    request_json: str
    base_revision: str | None
    result_revision: str | None
    result_json: str | None
    error_class: str | None
    error_message: str | None

    @property
    def replay_payload(self) -> dict[str, object] | None:
        if self.result_json is None:
            return None
        payload = json.loads(self.result_json)
        if not isinstance(payload, dict):
            return None
        return {str(key): value for key, value in payload.items()}


def _utcnow() -> str:
    return datetime.now(tz=UTC).replace(microsecond=0).isoformat().replace("+00:00", "Z")


def create_operation(connection: sqlite3.Connection, request: OperationRequest) -> OperationRecord:
    existing = connection.execute(
        "SELECT * FROM operations WHERE principal = ? AND idempotency_key = ?",
        (request.principal, request.idempotency_key),
    ).fetchone()
    if existing is not None:
        if existing["request_hash"] != request.request_hash:
            raise IdempotencyConflictError("idempotency key already used for a different request")
        return operation_from_row(existing)
    with connection:
        connection.execute(
            """
            INSERT INTO operations(
                op_id, idempotency_key, principal, client_instance_id, mcp_session_id,
                source_chat, tool_name, request_hash, state, request_json, created_at
            ) VALUES(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
            """,
            (
                request.op_id,
                request.idempotency_key,
                request.principal,
                request.client_instance_id,
                request.mcp_session_id,
                request.source_chat,
                request.tool_name,
                request.request_hash,
                OperationState.QUEUED,
                request.request_json,
                _utcnow(),
            ),
        )
    return get_operation(connection, request.op_id)


def get_operation_by_idempotency(
    connection: sqlite3.Connection, principal: str, idempotency_key: str
) -> OperationRecord | None:
    row = connection.execute(
        "SELECT * FROM operations WHERE principal = ? AND idempotency_key = ?",
        (principal, idempotency_key),
    ).fetchone()
    return operation_from_row(row) if row is not None else None


def get_operation(connection: sqlite3.Connection, op_id: str) -> OperationRecord:
    row = connection.execute("SELECT * FROM operations WHERE op_id = ?", (op_id,)).fetchone()
    if row is None:
        raise KeyError(op_id)
    return operation_from_row(row)


def list_interrupted_operations(connection: sqlite3.Connection) -> tuple[OperationRecord, ...]:
    rows = connection.execute(
        "SELECT * FROM operations WHERE state IN (?, ?, ?) ORDER BY created_at, op_id",
        (OperationState.QUEUED, OperationState.RUNNING, OperationState.RECOVERING),
    ).fetchall()
    return tuple(operation_from_row(row) for row in rows)


def mark_operation_running(
    connection: sqlite3.Connection, op_id: str, *, base_revision: str
) -> OperationRecord:
    return _update_operation(
        connection,
        op_id,
        state=OperationState.RUNNING,
        base_revision=base_revision,
        started_at=_utcnow(),
    )


def mark_operation_succeeded(
    connection: sqlite3.Connection,
    op_id: str,
    *,
    result_revision: str,
    result: dict[str, object],
) -> OperationRecord:
    return _update_operation(
        connection,
        op_id,
        state=OperationState.SUCCEEDED,
        result_revision=result_revision,
        result_json=json.dumps(result, sort_keys=True),
        error_class=None,
        error_message=None,
        finished_at=_utcnow(),
    )


def mark_operation_conflict(
    connection: sqlite3.Connection, op_id: str, *, error_message: str
) -> OperationRecord:
    return _update_operation(
        connection,
        op_id,
        state=OperationState.CONFLICT,
        error_class="conflict",
        error_message=error_message,
        finished_at=_utcnow(),
    )


def mark_operation_failed(
    connection: sqlite3.Connection, op_id: str, *, error_class: str, error_message: str
) -> OperationRecord:
    return _update_operation(
        connection,
        op_id,
        state=OperationState.FAILED,
        error_class=error_class,
        error_message=error_message,
        finished_at=_utcnow(),
    )


def _update_operation(
    connection: sqlite3.Connection, op_id: str, **values: object
) -> OperationRecord:
    assignments = ", ".join(f"{column} = ?" for column in values)
    parameters = [*values.values(), op_id]
    with connection:
        connection.execute(f"UPDATE operations SET {assignments} WHERE op_id = ?", parameters)
    return get_operation(connection, op_id)


def operation_from_row(row: sqlite3.Row) -> OperationRecord:
    return OperationRecord(
        op_id=row["op_id"],
        principal=row["principal"],
        idempotency_key=row["idempotency_key"],
        tool_name=row["tool_name"],
        request_hash=row["request_hash"],
        state=OperationState(row["state"]),
        request_json=row["request_json"],
        base_revision=row["base_revision"],
        result_revision=row["result_revision"],
        result_json=row["result_json"],
        error_class=row["error_class"],
        error_message=row["error_message"],
    )

File: go/control/operations.go
package control

import (
	"context"
	"crypto/sha256"
	"database/sql"
	"errors"
	"fmt"
	"time"

	"github.com/rcarmo/memento/go/internal/pyjson"
)

type OperationState string

const (
	Queued     OperationState = "queued"
	Running    OperationState = "running"
	Succeeded  OperationState = "succeeded"
	Failed     OperationState = "failed"
	Conflict   OperationState = "conflict"
	Recovering OperationState = "recovering"
)

type IdempotencyConflictError struct{}

func (*IdempotencyConflictError) Error() string {
	return "idempotency key already used for a different request"
}

type OperationNotFoundError struct{ OpID string }

func (e *OperationNotFoundError) Error() string { return "unknown operation: " + e.OpID }

type OperationRequest struct {
	OpID             string  `json:"op_id"`
	Principal        string  `json:"principal"`
	IdempotencyKey   string  `json:"idempotency_key"`
	ToolName         string  `json:"tool_name"`
	RequestJSON      string  `json:"request_json"`
	ClientInstanceID *string `json:"client_instance_id"`
	MCPSessionID     *string `json:"mcp_session_id"`
	SourceChat       *string `json:"source_chat"`
}

func (r OperationRequest) RequestHash() string {
	return fmt.Sprintf("%x", sha256.Sum256([]byte(r.RequestJSON)))
}

type OperationRecord struct {
	OpID           string         `json:"op_id"`
	Principal      string         `json:"principal"`
	IdempotencyKey string         `json:"idempotency_key"`
	ToolName       string         `json:"tool_name"`
	RequestHash    string         `json:"request_hash"`
	State          OperationState `json:"state"`
	RequestJSON    string         `json:"request_json"`
	BaseRevision   *string        `json:"base_revision"`
	ResultRevision *string        `json:"result_revision"`
	ResultJSON     *string        `json:"result_json"`
	ErrorClass     *string        `json:"error_class"`
	ErrorMessage   *string        `json:"error_message"`
}

func (r OperationRecord) ReplayPayload() (map[string]any, error) {
	if r.ResultJSON == nil {
		return nil, nil
	}
	value, err := pyjson.Parse(*r.ResultJSON)
	if err != nil {
		return nil, err
	}
	payload, _ := value.(map[string]any)
	return payload, nil
}

// Operations preserves the reference state transitions, including blind updates
// and retrying failed rows. Identity is the exact request JSON hash, scoped by
// principal/key. Service authorisation must occur before calling these methods.
// Separate handles may race a SELECT/INSERT and get a uniqueness error, just as
// in Python; transaction serialization is supplied by the transaction manager.
type Operations struct {
	DB  *sql.DB
	Now func() time.Time
}

func (o Operations) now() string {
	if o.Now != nil {
		return o.Now().UTC().Truncate(time.Second).Format(time.RFC3339)
	}
	return time.Now().UTC().Truncate(time.Second).Format(time.RFC3339)
}

const operationColumns = "op_id,principal,idempotency_key,tool_name,request_hash,state,request_json,base_revision,result_revision,result_json,error_class,error_message"

func (o Operations) Create(ctx context.Context, request OperationRequest) (OperationRecord, error) {
	existing, err := o.ByIdempotency(ctx, request.Principal, request.IdempotencyKey)
	if err != nil {
		return OperationRecord{}, err
	}
	if existing != nil {
		if existing.RequestHash != request.RequestHash() {
			return OperationRecord{}, &IdempotencyConflictError{}
		}
		return *existing, nil
	}
	_, err = o.DB.ExecContext(ctx, `INSERT INTO operations(op_id,idempotency_key,principal,client_instance_id,mcp_session_id,source_chat,tool_name,request_hash,state,request_json,created_at) VALUES(?,?,?,?,?,?,?,?,?,?,?)`, request.OpID, request.IdempotencyKey, request.Principal, request.ClientInstanceID, request.MCPSessionID, request.SourceChat, request.ToolName, request.RequestHash(), Queued, request.RequestJSON, o.now())
	if err != nil {
		return OperationRecord{}, err
	}
	return o.Get(ctx, request.OpID)
}
func (o Operations) Get(ctx context.Context, opID string) (OperationRecord, error) {
	record, err := scanOperation(o.DB.QueryRowContext(ctx, "SELECT "+operationColumns+" FROM operations WHERE op_id = ?", opID))
	if errors.Is(err, sql.ErrNoRows) {
		return OperationRecord{}, &OperationNotFoundError{OpID: opID}
	}
	return record, err
}
func (o Operations) ByIdempotency(ctx context.Context, principal, key string) (*OperationRecord, error) {
	record, err := scanOperation(o.DB.QueryRowContext(ctx, "SELECT "+operationColumns+" FROM operations WHERE principal = ? AND idempotency_key = ?", principal, key))
	if errors.Is(err, sql.ErrNoRows) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	return &record, nil
}
func (o Operations) Interrupted(ctx context.Context) ([]OperationRecord, error) {
	rows, err := o.DB.QueryContext(ctx, "SELECT "+operationColumns+" FROM operations WHERE state IN (?, ?, ?) ORDER BY created_at, op_id", Queued, Running, Recovering)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	return scanOperations(rows)
}

type operationRows interface {
	Next() bool
	Scan(...any) error
	Err() error
}

func scanOperations(rows operationRows) ([]OperationRecord, error) {
	out := []OperationRecord{}
	for rows.Next() {
		record, err := scanOperation(rows)
		if err != nil {
			return nil, err
		}
		out = append(out, record)
	}
	return out, rows.Err()
}

type rowScanner interface{ Scan(...any) error }

func scanOperation(row rowScanner) (OperationRecord, error) {
	var record OperationRecord
	if err := row.Scan(&record.OpID, &record.Principal, &record.IdempotencyKey, &record.ToolName, &record.RequestHash, &record.State, &record.RequestJSON, &record.BaseRevision, &record.ResultRevision, &record.ResultJSON, &record.ErrorClass, &record.ErrorMessage); err != nil {
		return OperationRecord{}, err
	}
	switch record.State {
	case Queued, Running, Succeeded, Failed, Conflict, Recovering:
		return record, nil
	default:
		return OperationRecord{}, fmt.Errorf("invalid operation state: %s", record.State)
	}
}
func (o Operations) MarkRunning(ctx context.Context, opID, base string) (OperationRecord, error) {
	return o.update(ctx, opID, "state=?, base_revision=?, started_at=?", Running, base, o.now())
}
func (o Operations) MarkSucceeded(ctx context.Context, opID, revision string, result map[string]any) (OperationRecord, error) {
	value, err := pyjson.Dumps(result)
	if err != nil {
		return OperationRecord{}, err
	}
	return o.update(ctx, opID, "state=?, result_revision=?, result_json=?, error_class=?, error_message=?, finished_at=?", Succeeded, revision, value, nil, nil, o.now())
}
func (o Operations) MarkConflict(ctx context.Context, opID, message string) (OperationRecord, error) {
	return o.update(ctx, opID, "state=?, error_class=?, error_message=?, finished_at=?", Conflict, "conflict", message, o.now())
}
func (o Operations) MarkFailed(ctx context.Context, opID, errorClass, message string) (OperationRecord, error) {
	return o.update(ctx, opID, "state=?, error_class=?, error_message=?, finished_at=?", Failed, errorClass, message, o.now())
}
func (o Operations) update(ctx context.Context, opID, assignments string, values ...any) (OperationRecord, error) {
	values = append(values, opID)
	if _, err := o.DB.ExecContext(ctx, "UPDATE operations SET "+assignments+" WHERE op_id=?", values...); err != nil {
		return OperationRecord{}, err
	}
	return o.Get(ctx, opID)
}
